spec/services/charts/annotate_spec.rb

```diff
@@ -1,250 +1,791 @@
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Awst 2022')).to eq(Date.parse('01/08/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Medi 2022')).to eq(Date.parse('01/09/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Hyd 2022')).to eq(Date.parse('01/10/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Tach 2022')).to eq(Date.parse('01/11/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Rhag 2022')).to eq(Date.parse('01/12/2022'))
-        I18n.locale = 'en'
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Jan 2022')).to eq(Date.parse('01/01/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Feb 2022')).to eq(Date.parse('01/02/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Mar 2022')).to eq(Date.parse('01/03/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Apr 2022')).to eq(Date.parse('01/04/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 May 2022')).to eq(Date.parse('01/05/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Jun 2022')).to eq(Date.parse('01/06/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Jul 2022')).to eq(Date.parse('01/07/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Aug 2022')).to eq(Date.parse('01/08/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Sep 2022')).to eq(Date.parse('01/09/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Oct 2022')).to eq(Date.parse('01/10/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Nov 2022')).to eq(Date.parse('01/11/2022'))
-        expect(Charts::Annotate.new(school).send(:date_for, '01 Dec 2022')).to eq(Date.parse('01/12/2022'))
+    context 'with no intervention or activity observations' do
+      it 'returns no annotations' do
+        expect(subject_multi_fuel.annotate_weekly(x_axis_categories)).to be_empty
       end
     end
 
-    describe 'annotating weekly charts' do
-
-      subject { Charts::Annotate.new(school).annotate_weekly(x_axis_categories)}
+    context 'show on charts' do
+      let!(:intervention_observation_multi_fuel) { create(:observation, :intervention, school: school, at: Date.new(2018, 6, 24), intervention_type: multi_fuel_intervention) }
+      let!(:activity_observation_multi_fuel) { create(:observation, :activity, school: school, at: Date.new(2018, 7, 8), activity: activity_multi_fuel) }
 
-      let(:x_axis_categories){ [
-        "24 Jun 2018",
-        "01 Jul 2018",
-        "08 Jul 2018",
-        "15 Jul 2018"
-      ]}
+      it 'is set true' do
+        multi_fuel_intervention.update(show_on_charts: true)
+        activity_type_multi_fuel.update(show_on_charts: true)
+        expect(subject_multi_fuel.annotate_weekly(x_axis_categories)).to eq(
+          [
```

Comment:
Testing for large data structures like this in output can make tests difficult to read / debug / maintain. I am not sure how I would do it exactly in this case, but I'd try and replace this one test with multiple tests to check the output contains certain items (to prove it's behaving as expected). e.g. `it "has two annotations"`. `it "has an intervention observation"` etc. Maybe one to think about another time and probably would need shared examples and further contexts to achieve this. This would also have the side effect of describing what the desired output looked like without having to infer it from looking at the data structure.

Reply:
I agree that's probably good practice but I think in this case (and also where we're doing similar elsewhere) everything in the array is relevant to correct behaviour (e.g. `Observation` dates being pinned to the nearest week date, the icon type etc).

Reply:
Thanks @iant. I'm always one for descriptive short tests where possible (not saying I always do it myself either), unless the test describes itself. Which it sounds like it does in this case, so, ok to leave as-is.